cfme/common/provider_views.py

```diff
@@ -223,13 +226,48 @@ class ProviderToolBar(View):
     view_selector = View.nested(ItemsToolBarViewSelector)
 
 
-class ProviderEntities(View):
+class ProviderItems(View):
     """
     should represent the view with different items like providers
     """
     title = Text('//div[@id="main-content"]//h1')
     search = View.nested(Search)
-    # todo: in progress
+    _quadicons = '//tr[./td/div[@class="quadicon"]]/following-sibling::tr/td/a'
+    _listitems = Table(locator='//div[@id="list_grid"]/table')
+
+    def _get_item_names(self):
+        if self.parent.toolbar.view_selector.selected == 'List View':
+            return [row.name.text for row in self._listitems.rows()]
+        else:
+            br = self.browser
+            return [br.get_attribute('title', el) for el in br.elements(self._quadicons)]
+
+    def get_all(self, surf_pages=False):
+        if not surf_pages:
+            return [ProviderItem(parent=self, name=name) for name in self._get_item_names()]
+        else:
+            items = []
+            for _ in self.parent.paginator.pages():
+                items.extend([ProviderItem(parent=self, name=name)
+                              for name in self._get_item_names()])
+            return items
+
+    def get_items(self, by_name=None, surf_pages=False):
+        items = self.get_all(surf_pages)
+        remaining_items = []
+        for item in items:
+            if by_name and by_name in item.name:
```

Comment:
Consider something like this, where you're only running this loop if necessary, instead of checking by_name down here.  Pretty sure this gets you the same data structure.

```python
items = self.get_all(surf_pages)
if not by_name:
    return items
elseif by_name:
    return [item for item in items if by_name in item.name]